storage/reader.go

```diff
@@ -392,6 +574,72 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
+	if r.stream == nil {
+		return 0, io.EOF
+	}
+
+	// The entire object has been read by this reader, return EOF.
+	if r.size != 0 && r.size == r.seen {
+		return 0, io.EOF
+	}
+
+	n := copy(p, r.leftovers)
+	r.seen += int64(n)
+	r.leftovers = r.leftovers[n:]
+
+	for len(p[n:]) > 0 {
+		msg, err := r.stream.Recv()
+		if err != nil {
+			if err == io.EOF && r.seen == r.size {
+				// Let the next call to readWithGRPC return EOF.
+				return n, nil
+			}
+			// This prevents reopening of the stream when the error is not
+			// retryable.
+			if !shouldRetry(err) {
+				return n, err
+			}
+
+			// Save the original error to return instead of the reinit error.
+			e := err
+
+			// Close existing stream and initialize new stream with updated
+			// offset.
+			r.closeStream()
+
+			// This will immediately attempt to reopen the stream, but will
+			// backoff if further attempts are necessary.
+			var res *readStreamResponse
+			res, r.cancelStream, err = r.reopenWithGRPC(r.seen)
+			if err != nil {
+				// Cannot reinstantiate the stream so return the original error.
```

Comment:
Should we wrap these perhaps? Is there any use in preserving the reopen error?

Reply:
I can wrap it, yes. 

As for the reopen error, I'm not sure. Attempting to `Recv` from the stream should've produced the "real" error  (e.g. BadRequest), so I don't want to lose that. But if we attempt to reopen and _that_ fails, maybe that is useful to know as well. Should be return both? `fmt.errorf("reading stream failed: %s and stream reopen failed: %s", e, err)`

Reply:
No longer necessary due to refactoring of Recv with retry. Spoke offline.